sdk/mobile-center/src/main/java/com/microsoft/azure/mobile/utils/crypto/CryptoUtils.java

```diff
@@ -0,0 +1,493 @@
+        if (data == null)
+            return null;
+        try {
+
+            /* Get preferred crypto handler. */
+            CryptoHandlerEntry handlerEntry = mCryptoHandlers.values().iterator().next();
+            CryptoHandler handler = handlerEntry.mCryptoHandler;
+            try {
+
+                /* Attempt encryption. */
+                KeyStore.Entry keyStoreEntry = getKeyStoreEntry(handlerEntry);
+                byte[] encryptedBytes = handler.encrypt(mCryptoFactory, mApiLevel, keyStoreEntry, data.getBytes(CHARSET));
+                String encryptedString = Base64.encodeToString(encryptedBytes, Base64.DEFAULT);
+
+                /*
+                 * Store algorithm for crypto agility alongside the data.
+                 * We also use that information in decrypt in case of firmware/sdk upgrade.
+                 */
+                return handler.getAlgorithm() + ALGORITHM_DATA_SEPARATOR + encryptedString;
+            } catch (InvalidKeyException e) {
+
+                /* When key expires, switch to another alias. */
+                MobileCenterLog.debug(LOG_TAG, "Alias expired: " + handlerEntry.mAliasIndex);
+                handlerEntry.mAliasIndex ^= 1;
+                String newAlias = getAlias(handler, handlerEntry.mAliasIndex);
+
+                /* If this is the second time we switch, we delete the previous key. */
+                if (mKeyStore.containsAlias(newAlias)) {
+                    MobileCenterLog.debug(LOG_TAG, "Deleting alias: " + newAlias);
+                    mKeyStore.deleteEntry(newAlias);
+                }
+
+                /* Generate new key. */
+                MobileCenterLog.debug(LOG_TAG, "Creating alias: " + newAlias);
+                handler.generateKey(mCryptoFactory, mApiLevel, newAlias, mContext);
+
+                /* And encrypt using that new key. */
+                return encrypt(data);
+            }
+        } catch (Exception e) {
+
+            /* Return data as is. */
+            MobileCenterLog.error(LOG_TAG, "Failed to encrypt data.");
+            return data;
+        }
+    }
+
+    /**
+     * Decrypt data.
+     *
+     * @param data data to decrypt.
+     * @return decrypted data.
+     */
+    @NonNull
+    public DecryptedData decrypt(@Nullable String data) {
+
+        /* Handle null for convenience. */
+        if (data == null)
+            return new DecryptedData(null, null);
+
+        /* Guess what algorithm was used in case the data was encrypted using an old SDK or old firmware. */
+        String[] dataSplit = data.split(ALGORITHM_DATA_SEPARATOR);
+        CryptoHandlerEntry handlerEntry = dataSplit.length == 2 ? mCryptoHandlers.get(dataSplit[0]) : null;
+        CryptoHandler cryptoHandler = handlerEntry == null ? null : handlerEntry.mCryptoHandler;
+        try {
+            if (cryptoHandler == null) {
+                throw new IllegalStateException("Could not find crypto handler that was used for the specified data.");
```

Comment:
The message is not used in catch statement.
It would be good to check before getting into try statement and log a proper message for this case.